tests/compilation_tests/grid_test.py

```diff
@@ -11,129 +11,212 @@
+       top_right_node = g.nodes[g.get_num_rows()-1][g.get_num_cols()-1]
+       self.assertLessEqual(top_right_node.get_m_coords()[
+                            0], x_range, "The meters grid shouldn't be larger than the lat bounds")
+       self.assertLessEqual(top_right_node.get_m_coords()[
+                            1], y_range, "The meters grid shouldn't be larger than the long bounds")
+
+   def test_inside(self):
+       lat_min, lat_max, long_min, long_max = 42.444250, 42.444599, -76.483682, -76.483276
+       g = Grid(lat_min, lat_max, long_min, long_max)
+       self.assertTrue(g.is_inside_triangle(2, 2, 3, 3, 1, 3, 2,
+                       2.5), "this point is inside the triangle")
+       self.assertFalse(g.is_inside_triangle(2, 2, 3, 3, 1, 3,
+                        10, 10), "this point is outside the triangle")
+       self.assertTrue(g.is_inside_triangle(2, 2, 3, 3, 1, 3, 2,
+                       2), "this point is inside the triangle")
+
+   def test_is_on_border_triangle(self):
+       count = 0
+       lat_min, lat_max, long_min, long_max = 42.444250, 42.444599, -76.483682, -76.483276
+       g = Grid(lat_min, lat_max, long_min, long_max)
+      
+       g.activate_triangle(1,1,5,5,9,1)
+       g.find_border_nodes()
+
+       count = 0
+       rows = g.nodes.shape[0]
+       cols = g.nodes.shape[1]
+       for row in range(rows):
+           for col in range(cols):
+               node = g.nodes[row][col]
+               if node.is_active_node() and g.is_on_border(row, col, rows-1, cols-1):
+                   count += 1
+
+       self.assertEqual(len(g.border_nodes), count)
+
+   def test_is_on_border_rectangle(self):
+       lat_min, lat_max, long_min, long_max = 42.444250, 42.444599, -76.483682, -76.483276
+       g = Grid(lat_min, lat_max, long_min, long_max)
+      
+       rows = g.nodes.shape[0]
+       cols = g.nodes.shape[1]
+
+       g.activate_rectangle(0, 0, rows, cols)
+       g.find_border_nodes()
+
+       count = 0
+       for row in range(rows):
+           for col in range(cols):
+               node = g.nodes[row][col]
+               if node.is_active_node() and g.is_on_border(row, col, rows-1, cols-1):
+                   count += 1
+                  
+       self.assertEqual(len(g.border_nodes), count)
+
+   def test_is_on_border_circle(self):   
+       lat_min, lat_max, long_min, long_max = 42.444250, 42.444599, -76.483682, -76.483276
+       g = Grid(lat_min, lat_max, long_min, long_max)
+      
+       rows = g.nodes.shape[0]
+       cols = g.nodes.shape[1]
+
+       g.activate_circle(0, 0, 5)
+       g.find_border_nodes()
+
+       count = 0
+       for row in range(rows):
+           for col in range(cols):
+               node = g.nodes[row][col]
+               if node.is_active_node() and g.is_on_border(row, col, rows-1, cols-1):
+                   count += 1
+                  
+       self.assertEqual(len(g.border_nodes), count)
```

Comment:
For test_is_on_border_triangle, test_is_on_border_rectangle, test_is_on_border_circle, to avoid code repetition, could you combine these three under 1 function?

Reply:
I have noticed that your test basically copies the code for is_on_border_node and then asserts equals. This type of test will always pass and will probably not find bugs. Maybe can you try creating a test case where you activate a small shape and manually check if the border nodes are equal to what the function finds?